### PR TITLE
[BugFix] Fix date type as partition value when inserting iceberg table

### PR DESCRIPTION
**Pull request by stephen-shelby**
Fixes #issue
We need to get date string like "YYYY-MM-dd" from Datum. But the value in the DateColumn is JuLian of DateValue. It's an int_32 type. So we need to transform it to date string. Additionally, we should restrict unsupported data types from being used as partition columns.

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 1]
```diff
@@ -113,7 +112,14 @@ Status IcebergTableSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr&
         const std::vector<std::string>& partition_column_names = _iceberg_table->partition_column_names();
         std::vector<std::string> partition_column_values;
         for (const ColumnPtr& column : partitions_columns) {
-            partition_column_values.emplace_back(_value_to_string(column));
+            std::string partition_value;
+            if (column->is_nullable()) {
+                const auto& nullable_column = std::static_pointer_cast<NullableColumn>(column);
```
Comment (zombee0) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:117]: This is dangerous, const-null-column is_nullable(), then cast will crash

Comment (letian-jiang) [reply be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:117]: > This is dangerous, const-null-column is_nullable(), then cast will crash

Is there any util function to use in this case?

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 2]
```diff
@@ -96,7 +94,8 @@ Status IcebergTableSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr&
         if (_partition_writers.empty()) {
             tableInfo.partition_location = _iceberg_table_data_location;
             auto writer = std::make_unique<RollingAsyncParquetWriter>(tableInfo, _output_expr, _common_metrics.get(),
-                                                                      add_iceberg_commit_info, state, _driver_sequence);
+                                                                      IcebergTableSinkOperator::add_iceberg_commit_info,
```
Comment (letian-jiang) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:97]: `IcebergTableSinkOperator::` seems redundant

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 3]
```diff
@@ -124,7 +130,8 @@ Status IcebergTableSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr&
         if (partition_writer == _partition_writers.end()) {
             tableInfo.partition_location = partition_location;
             auto writer = std::make_unique<RollingAsyncParquetWriter>(tableInfo, _output_expr, _common_metrics.get(),
-                                                                      add_iceberg_commit_info, state, _driver_sequence);
+                                                                      IcebergTableSinkOperator::add_iceberg_commit_info,
```
Comment (letian-jiang) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:133]: ditto

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp (hunk 1 above)
Comment (letian-jiang) [reply be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:117]: @zombee0 Is there any util function to use in this case?

Comment (stephen-shelby) [reply be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp]: I have banned the null column.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by letian-jiang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|5|100.00%|[]|

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 4]
```diff
@@ -333,4 +319,35 @@ static void add_iceberg_commit_info(starrocks::parquet::AsyncFileWriter* writer,
...
+    auto v = column->get(0);
+
+    if (column->is_date()) {
+        partition_value = v.get_date().to_string();
+        return Status::OK();
+    }
+
+    bool not_support = false;
+    v.visit([&](auto& variant) {
+        std::visit(
```
Comment (dirtysalt) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:332]: why use std::visit?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|5|100.00%|[]|

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp (hunk 4 above)
Comment (stephen-shelby) [reply be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:326]: communicated privately

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|5|100.00%|[]|

**Review by dirtysalt — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|5|100.00%|[]|

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**